sycl/include/sycl/reduction.hpp

```diff
@@ -1200,13 +1200,18 @@ template <> struct NDRangeReduction<reduction::strategy::range_basic> {
         // Reduce each result separately
         // TODO: Opportunity to parallelize across elements
         for (int E = 0; E < NElements; ++E) {
-          auto LocalSum = Identity;
-          for (size_t I = LID; I < NWorkGroups; I += WGSize)
-            LocalSum = BOp(LocalSum, PartialSums[I * NElements + E]);
+          // The last work-group may not have enough work for all its items.
+          size_t RemainingWorkSize = std::min(NWorkGroups, WGSize);
 
-          LocalReds[LID] = LocalSum;
+          if (LID < RemainingWorkSize) {
+            auto LocalSum = PartialSums[LID * NElements + E];
+            for (size_t I = LID + WGSize; I < NWorkGroups; I += WGSize)
+              LocalSum = BOp(LocalSum, PartialSums[I * NElements + E]);
 
-          doTreeReduction(WGSize, LID, LocalReds, BOp,
+            LocalReds[LID] = LocalSum;
+          }
```

Comment:
I'd suggest moving that inside `doTreeReduction`, possibly adding an extra `template <bool SizeFitsWGSize = false>` for the purposes of performance optimizations.